modules/reindex/src/test/java/org/elasticsearch/index/reindex/remote/RemoteScrollableHitSourceTests.java

```diff
@@ -0,0 +1,381 @@
+            assertEquals(14, failure.getColumnNumber());
+            called.set(true);
+        };
+        sourceWithMockedRemoteCall("request_failure.json").doStart(checkResponse);
+        assertTrue(called.get());
+        called.set(false);
+        sourceWithMockedRemoteCall("request_failure.json").doStartNextScroll("scroll", timeValueMillis(0), checkResponse);
+        assertTrue(called.get());
+    }
+
+    public void testRetryAndSucceed() throws Exception {
+        AtomicBoolean called = new AtomicBoolean();
+        Consumer<InternalResponse> checkResponse = r -> {
+            assertThat(r.getFailures(), hasSize(0));
+            called.set(true);
+        };
+        retriesAllowed = between(1, Integer.MAX_VALUE);
+        sourceWithMockedRemoteCall("fail:rejection.json", "start_ok.json").doStart(checkResponse);
+        assertTrue(called.get());
+        assertEquals(1, retries);
+        retries = 0;
+        called.set(false);
+        sourceWithMockedRemoteCall("fail:rejection.json", "scroll_ok.json").doStartNextScroll("scroll", timeValueMillis(0),
+                checkResponse);
+        assertTrue(called.get());
+        assertEquals(1, retries);
+    }
+
+    public void testRetryUntilYouRunOutOfTries() throws Exception {
+        AtomicBoolean called = new AtomicBoolean();
+        Consumer<InternalResponse> checkResponse = r -> called.set(true);
+        retriesAllowed = between(0, 10);
+        String[] paths = new String[retriesAllowed + 2];
+        for (int i = 0; i < retriesAllowed + 2; i++) {
+            paths[i] = "fail:rejection.json";
+        }
+        RuntimeException e = expectThrows(RuntimeException.class, () -> sourceWithMockedRemoteCall(paths).doStart(checkResponse));
+        assertEquals("failed", e.getMessage());
+        assertFalse(called.get());
+        assertEquals(retriesAllowed, retries);
+        retries = 0;
+        e = expectThrows(RuntimeException.class,
+                () -> sourceWithMockedRemoteCall(paths).doStartNextScroll("scroll", timeValueMillis(0), checkResponse));
+        assertEquals("failed", e.getMessage());
+        assertFalse(called.get());
+        assertEquals(retriesAllowed, retries);
+    }
+
+    private RemoteScrollableHitSource sourceWithMockedRemoteCall(String... paths) throws Exception {
+        return sourceWithMockedRemoteCall(true, paths);
+    }
+
+    /**
+     * Creates a hit source that doesn't make the remote request and instead returns data from some files. Also requests are always returned
+     * synchronously rather than asynchronously.
+     */
+    private RemoteScrollableHitSource sourceWithMockedRemoteCall(boolean mockRemoteVersion, String... paths) throws Exception {
+        URL[] resources = new URL[paths.length];
+        for (int i = 0; i < paths.length; i++) {
+            resources[i] = Thread.currentThread().getContextClassLoader().getResource("responses/" + paths[i].replace("fail:", ""));
```

Comment:
The [javadoc](https://docs.oracle.com/javase/7/docs/api/java/lang/ClassLoader.html#getResource(java.lang.String)) says: 

```
The name of a resource is a '/'-separated path name that identifies the resource.
```

But I admit I haven't tried it on windows and now that I look at it I wouldn't be surprised if it didn't work there. I'm 80% sure it'll just work on windows.